LBH.AdultSocialCare.Api/V1/UseCase/Common/Concrete/UpdateCarePackageUseCase.cs

```diff
@@ -0,0 +1,42 @@
+using AutoMapper;
+using LBH.AdultSocialCare.Api.V1.Boundary.Common.Response;
+using LBH.AdultSocialCare.Api.V1.Domain.Common;
+using LBH.AdultSocialCare.Api.V1.Factories;
+using LBH.AdultSocialCare.Api.V1.Gateways;
+using LBH.AdultSocialCare.Api.V1.Gateways.Common.Interfaces;
+using LBH.AdultSocialCare.Api.V1.UseCase.Common.Interfaces;
+using System;
+using System.Threading.Tasks;
+
+namespace LBH.AdultSocialCare.Api.V1.UseCase.Common.Concrete
+{
+    public class UpdateCarePackageUseCase : IUpdateCarePackageUseCase
+    {
+        private readonly IMapper _mapper;
+        private readonly IDatabaseManager _dbManager;
+        private readonly ICarePackageGateway _carePackageGateway;
+        private readonly ICarePackageSettingsGateway _carePackageSettings;
+
+        public UpdateCarePackageUseCase(IMapper mapper, IDatabaseManager dbManager, ICarePackageGateway carePackageGateway, ICarePackageSettingsGateway carePackageSettings)
+        {
+            _mapper = mapper;
+            _dbManager = dbManager;
+            _carePackageGateway = carePackageGateway;
+            _carePackageSettings = carePackageSettings;
+        }
+
+        public async Task<CarePackagePlainResponse> UpdateAsync(Guid carePackageId, CarePackageUpdateDomain carePackageUpdateDomain)
+        {
+            var package = await _carePackageGateway.CheckPackageExistsAsync(carePackageId, true);
```

Comment:
CheckSomethingExists implies method returning true/false
Consider using EnsureSomething. It's a typical Microsoft naming pattern for methods which get something and throw / create object if object doesn't exist. They use it everywhere, e.g. see a random source file https://github.com/microsoft/referencesource/blob/5697c29004a34d80acdaf5742d7e699022c64ecd/System.Data.Entity/System/Data/Objects/Internal/ObjectQueryState.cs#L276

Ensure in this context means "provide with warranty / guarantee"

Since null checking is out scope of gateway logic, consider using helper (just don't rename it EnsureHelper please :)) like

```
var package = Ensure.Exists(
    () => _carePackageGateway.GetPackageAsync(packageId), $"Care package {packageId} not found");

public static class Ensure
    {
        public static async Task<T> Exists<T>(Func<Task<T>> action, string message = "Entity is not found")
        {
            if (action is null) throw new ArgumentNullException(nameof(action));

            var entity = await action.Invoke();

            if (entity is null)
            {
                throw new ApiException(message, HttpStatusCode.NotFound);
            }

            return entity;
        }
    }
```

Reply:
Also I'm thinking about using gateway decorators like 

```
                    /* or AsNoTracking */
_gateway.EnsureExisting().AsReadOnly().GetPackagesAsync()
```

This approach is more complex, so will think about it at spare time..

Reply:
Another option is extension method like 

```
var package = await _carePackageGateway
    .GetPackageAsync(packageId)
    .EnsureExists($"Care package {packageId} not found");

public static class ObjectExtensions
{
    public static T EnsureExists<T>(this T obj, string message)
    {
        if (obj is null)
        {
            throw new ApiException(message, HttpStatusCode.NotFound);
        }

        return obj;
    }
}
```

Reply:
I vote the last option. Making that change right now